sql: implement virtual index lookup join

jordanlewis opened the pull request:
This commit adds the capability to plan inner lookup joins into virtual
tables with indexes. This should be a big boon to certain complex,
chained joins that tend to come up from ORMs querying database metadata.

Closes #48178.

Release note (performance improvement): the optimizer and execution
engine can now plan lookup joins into virtual indexes, avoiding full
scans against virtual tables.

jordanlewis committed:
sql: implement virtual index lookup join

This commit adds the capability to plan lookup joins into virtual
tables with indexes. This should be a big boon to certain complex,
chained joins that tend to come up from ORMs querying database metadata.

Release note (performance improvement): the optimizer and execution
engine can now plan lookup joins into virtual indexes, avoiding full
scans against virtual tables.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48226)
<!-- Reviewable:end -->

rohany commented:
:pogchamp:

jordanlewis commented:
This still needs a lot of tests. Also, I need to figure out how to adjust the optimizer costing to nearly always prefer lookup joins against virtual indexes. Right now, for the big query we're testing, it still plans hash/merge joins.

The query is:

```
SELECT
        t2.oid::REGCLASS AS to_table,
        a1.attname AS column,
        a2.attname AS primary_key,
        c.conname AS name,
        c.confupdtype AS on_update,
        c.confdeltype AS on_delete,
        c.convalidated AS valid
FROM
        pg_constraint AS c
        JOIN pg_class AS t1 ON c.conrelid = t1.oid
        JOIN pg_class AS t2 ON c.confrelid = t2.oid
        JOIN pg_attribute AS a1 ON
                        (a1.attnum = c.conkey[1]) AND (a1.attrelid = t1.oid)
        JOIN pg_attribute AS a2 ON
                        (a2.attnum = c.confkey[1]) AND (a2.attrelid = t2.oid)
        JOIN pg_namespace AS t3 ON c.connamespace = t3.oid
WHERE
        ((c.contype = 'f') AND (t1.oid = 'b'::regclass))
        AND (t3.nspname = ANY (current_schemas(false)))
ORDER BY
        c.conname
```

cc @rafiss

jordanlewis commented:
@RaduBerinde, how would you recommend I adjust costs here? Perhaps the ideal thing would be to put some basic statistics in - we should have row counts on `system.descriptor`, and then maybe we could mark each of the virtual indexes that look up into `system.descriptor` about that row count.

Or, should we just up-weight the cost of scans in virtual tables overall?

RaduBerinde reviewed (commented):
For virtual tables it would make sense to not distinguish costs between "random" access and sequential access. But I'd rather not complicate things too much. With no stats, we pretend they have 1000 rows, which should be plenty to prefer lookup joins in most cases (and we can increase it for vtables if need be). The query you pasted above looks like this with your change:
```
  sort
   └── project
        ├── inner-join (hash)
        │    ├── select
        │    │    ├── scan t3
        │    │    └── filters
        │    │         └── nspname = ANY current_schemas(false)
        │    ├── inner-join (lookup pg_attribute@pg_attribute_attrelid_idx)
        │    │    ├── lookup columns are key
        │    │    ├── project
        │    │    │    ├── inner-join (lookup pg_attribute@pg_attribute_attrelid_idx)
        │    │    │    │    ├── lookup columns are key
        │    │    │    │    ├── project
        │    │    │    │    │    ├── inner-join (lookup pg_class@pg_class_oid_idx)
        │    │    │    │    │    │    ├── lookup columns are key
        │    │    │    │    │    │    ├── inner-join (lookup pg_class@pg_class_oid_idx)
        │    │    │    │    │    │    │    ├── lookup columns are key
        │    │    │    │    │    │    │    ├── select
        │    │    │    │    │    │    │    │    ├── scan c
        │    │    │    │    │    │    │    │    └── filters
        │    │    │    │    │    │    │    │         ├── conrelid = 'b'::REGCLASS
        │    │    │    │    │    │    │    │         └── contype = 'f'
        │    │    │    │    │    │    │    └── filters (true)
        │    │    │    │    │    │    └── filters
        │    │    │    │    │    │         └── t1.oid = 'b'::REGCLASS
        │    │    │    │    │    └── projections
        │    │    │    │    │         └── conkey[1]
        │    │    │    │    └── filters
        │    │    │    │         ├── column108 = a1.attnum
        │    │    │    │         └── a1.attrelid = 'b'::REGCLASS
        │    │    │    └── projections
        │    │    │         └── confkey[1]
        │    │    └── filters
        │    │         └── column131 = a2.attnum
        │    └── filters
        │         └── connamespace = t3.oid
        └── projections
```

It's all lookup joins except one at the top (`pg_namespace` has no index).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-:-M6GxUdr-AHBNQtv0_rQ:b-n08sy8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rohany)



<!-- Sent from Reviewable.io -->

jordanlewis commented:
Huh, that's not the plan I got when I tried it. I don't understand.

rohany commented on pkg/sql/virtual_table.go:
```diff
@@ -179,3 +182,133 @@ func (n *virtualTableNode) Close(ctx context.Context) {
+			v.scratch.indexKeyDatums,
+			v.table.ColumnIdxMap(),
+			&idxConstraint,
+			v.vtableCols)
```
[nit]: paren on next line

rohany reviewed (commented):
No real comments on the SQL exec portion -- this is super slick!

jordanlewis commented:
@RaduBerinde would you mind taking a quick look at this guy? Thanks!

RaduBerinde reviewed (approved):
Sorry for the delay, I've been pretty swamped. This looks great, nice work!

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-:-M6qJ-NvBvfJ8KXPZ4hR:bdky92b)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[pkg/sql/opt_exec_factory.go, line 718 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-M6GzhMHEvHBKLFD6EU2:-M6GzhMHEvHBKLFD6EU3:b-2nsj1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d93f75085099b840e82c04106bdcff6969132ad4/pkg/sql/opt_exec_factory.go#L718)):*
> ```Go
> 		onCond = nil
> 	}
> 	var n exec.Node
> ```

[nit] can just do `n :=` below
___
*[pkg/sql/opt_exec_factory.go, line 692 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-M6qHxpxE3QSSmWhYO9g:-M6qHxpxE3QSSmWhYO9h:boel2iy) ([raw file](https://github.com/cockroachdb/cockroach/blob/96276f7ca57befe1d6914dc1246122a30f2ed079/pkg/sql/opt_exec_factory.go#L692)):*
> ```Go
> }
> 
> func (ef *execFactory) constructVirtualLookupJoin(
> ```

[nit] add a short comment or improve the naming to `constructVirtualTableLookupJoin`
___
*[pkg/sql/opt_exec_factory.go, line 707 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-M6qICjp8eVoQAfj1N3s:-M6qICjp8eVoQAfj1N3t:bfka8vw) ([raw file](https://github.com/cockroachdb/cockroach/blob/96276f7ca57befe1d6914dc1246122a30f2ed079/pkg/sql/opt_exec_factory.go#L707)):*
> ```Go
> 	}
> 	if len(eqCols) > 1 {
> 		panic("vtable indexes with more than one column aren't supported yet")
> ```

why not just return an AssertionFailedf error
___
*[pkg/sql/virtual_schema.go, line 307 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-M6qJGSt79whUzC0M8hC:-M6qJGSt79whUzC0M8hD:b-ib1ihz) ([raw file](https://github.com/cockroachdb/cockroach/blob/96276f7ca57befe1d6914dc1246122a30f2ed079/pkg/sql/virtual_schema.go#L307)):*
> ```Go
> 		return errors.AssertionFailedf("datum row count and column count differ: %d vs %d", r, c)
> 	}
> 	for i, col := range columns {
> ```

[nit] since this gets evaluated per row, maybe avoid the ResultColumn copy into `col`
___
*[pkg/sql/virtual_schema.go, line 320 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-M6qJRO82OScdE5zQjpN:-M6qJRO82OScdE5zQjpO:b-tcnkom) ([raw file](https://github.com/cockroachdb/cockroach/blob/96276f7ca57befe1d6914dc1246122a30f2ed079/pkg/sql/virtual_schema.go#L320)):*
> ```Go
> 	}
> 	return nil
> ```

[nit] extra line
___
*[pkg/sql/virtual_table.go, line 216 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-M6qJYi1BEakrg5K9201:-M6qJYi1BEakrg5K9202:bb5x8uv) ([raw file](https://github.com/cockroachdb/cockroach/blob/96276f7ca57befe1d6914dc1246122a30f2ed079/pkg/sql/virtual_table.go#L216)):*
> ```Go
> 	// lookupCols is the projection on vtableCols to apply.
> 	lookupCols exec.TableColumnOrdinalSet
> ```

[nit] In other nodes, we separated the "runtime" state into another struct (which can be embedded). I think it has been helpful especially when just looking at planning/opt-factory side code.
___
*[pkg/sql/walk.go, line 945 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-M6qJwmh2CB6lO0LlBhu:-M6qJwmh2CB6lO0LlBhv:bbz9hz7) ([raw file](https://github.com/cockroachdb/cockroach/blob/96276f7ca57befe1d6914dc1246122a30f2ed079/pkg/sql/walk.go#L945)):*
> ```Go
> 	reflect.TypeOf(&valuesNode{}):            "values",
> 	reflect.TypeOf(&virtualTableNode{}):      "virtual table values",
> 	reflect.TypeOf(&vTableLookupJoinNode{}):  "virtual table lookup",
> ```

[nit] `virtual table lookup join` or `virtual-table-lookup-join` (for consistency with `lookup-join`)
___
*[pkg/sql/logictest/testdata/logic_test/orms, line 75 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-M6qKdbp2vzByxXdoAuy:-M6qKdbp2vzByxXdoAuz:borpoix) ([raw file](https://github.com/cockroachdb/cockroach/blob/96276f7ca57befe1d6914dc1246122a30f2ed079/pkg/sql/logictest/testdata/logic_test/orms#L75)):*
> ```
> ----
> name              primary  unique  indkey  column_indexes  column_names  definition
> customers_id_idx  false    false   2       {1,2}           {name,id}     CREATE INDEX customers_id_idx ON test.public.customers USING btree (id ASC)
> ```

Isn't this flaky, at least in principle? May want to add `ORDER BY` to the array_aggs
___
*[pkg/sql/opt/xform/custom_funcs.go, line 1323 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-M6qKC-e4eBVSIKru63w:-M6qKC-e4eBVSIKru63x:b675xx0) ([raw file](https://github.com/cockroachdb/cockroach/blob/96276f7ca57befe1d6914dc1246122a30f2ed079/pkg/sql/opt/xform/custom_funcs.go#L1323)):*
> ```Go
> 	if md.Table(scanPrivate.Table).IsVirtualTable() {
> 		if joinType != opt.InnerJoinOp {
> 			return
> ```

It wouldn't be hard to support LeftJoin no? Not sure how useful though


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48226)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde and @rohany)

___
*[pkg/sql/opt_exec_factory.go, line 707 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-M6qICjp8eVoQAfj1N3s:-M6rUaDK5mDiIoGiCNNt:b-hja4ig) ([raw file](https://github.com/cockroachdb/cockroach/blob/96276f7ca57befe1d6914dc1246122a30f2ed079/pkg/sql/opt_exec_factory.go#L707)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

why not just return an AssertionFailedf error
</blockquote></details>

If this error fires it means the binary is basically corrupted, haha - if someone defines a vtable w/ composite index it should really blow up badly in tests rather than make a usable binary. I made this an assertion error though, i think we will fail with this same problem elsewhere loudly.
___
*[pkg/sql/virtual_table.go, line 274 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-M6Ae18zv8bB2VFPsxk7-r2-274:-M6rcgFt71RcL9CAzhgI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/55a4d540a890797371246f24a1d29fcc53fea9d8/pkg/sql/virtual_table.go#L274)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

[nit]: paren on next line
</blockquote></details>

Done.
___
*[pkg/sql/virtual_table.go, line 216 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-M6qJYi1BEakrg5K9201:-M6rcUtX4CErK8L-X10h:b-jakiz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/96276f7ca57befe1d6914dc1246122a30f2ed079/pkg/sql/virtual_table.go#L216)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] In other nodes, we separated the "runtime" state into another struct (which can be embedded). I think it has been helpful especially when just looking at planning/opt-factory side code.
</blockquote></details>

Good point, thanks. Done.
___
*[pkg/sql/logictest/testdata/logic_test/orms, line 75 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-M6qKdbp2vzByxXdoAuy:-M6rcnV1BP0MS1xxQf7-:bah2p1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/96276f7ca57befe1d6914dc1246122a30f2ed079/pkg/sql/logictest/testdata/logic_test/orms#L75)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Isn't this flaky, at least in principle? May want to add `ORDER BY` to the array_aggs
</blockquote></details>

I want to keep this query as-is, it's a copy of a real world orm query. The fact that this changed is a little sad but I think ok.
___
*[pkg/sql/opt/xform/custom_funcs.go, line 1323 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-M6qKC-e4eBVSIKru63w:-M6rnDVSEOSjY5qI91Ja:b-n6zwrc) ([raw file](https://github.com/cockroachdb/cockroach/blob/96276f7ca57befe1d6914dc1246122a30f2ed079/pkg/sql/opt/xform/custom_funcs.go#L1323)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It wouldn't be hard to support LeftJoin no? Not sure how useful though
</blockquote></details>

Fair, was just being lazy. Done, and added some tests (and a giant ORM query that uses outer lookup join, yay)


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-:-M6vRUAxBL2IntDyqZwT:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @rohany)

___
*[pkg/sql/opt_exec_factory.go, line 707 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-M6qICjp8eVoQAfj1N3s:-M6vQCLt93L9VK1e-I27:bfqv2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/96276f7ca57befe1d6914dc1246122a30f2ed079/pkg/sql/opt_exec_factory.go#L707)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

If this error fires it means the binary is basically corrupted, haha - if someone defines a vtable w/ composite index it should really blow up badly in tests rather than make a usable binary. I made this an assertion error though, i think we will fail with this same problem elsewhere loudly.
</blockquote></details>

Thanks! There could in principle be bugs anywhere along the way (in the catalog, in the optimizer, in the exec builder, in the exec factory). An assertion error should be no less visible than a panic, especially in tests.
___
*[pkg/sql/logictest/testdata/logic_test/orms, line 75 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-M6qKdbp2vzByxXdoAuy:-M6vQjxCEOAcMUyrLvgE:bx2moum) ([raw file](https://github.com/cockroachdb/cockroach/blob/96276f7ca57befe1d6914dc1246122a30f2ed079/pkg/sql/logictest/testdata/logic_test/orms#L75)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I want to keep this query as-is, it's a copy of a real world orm query. The fact that this changed is a little sad but I think ok.
</blockquote></details>

Fine with me, if you think the output is deterministic (I don't see why not since it's all executed on a single node).
___
*[pkg/sql/opt/xform/custom_funcs.go, line 1323 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48226#-M6qKC-e4eBVSIKru63w:-M6vRMNl-Z75DJr4XS0a:b903krw) ([raw file](https://github.com/cockroachdb/cockroach/blob/96276f7ca57befe1d6914dc1246122a30f2ed079/pkg/sql/opt/xform/custom_funcs.go#L1323)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Fair, was just being lazy. Done, and added some tests (and a giant ORM query that uses outer lookup join, yay)
</blockquote></details>

Awesome!!j


<!-- Sent from Reviewable.io -->

jordanlewis committed:
sql: bugfix to pg_aggregate

Commit 4047d45256b8cc7291d0167916115066d4f09f90 introduced a bug which
caused reads from pg_aggregate to mistakenly change the 0 OID to have a
"-" name alias, causing some inconsistent behavior in system tables.

Release note: None

jordanlewis commented:
bors r+

jordanlewis commented:
bors r+

jordanlewis commented:
Grr, flaky test "test missing log output".

bors r+

jordanlewis commented:
One more try:

bors r+

jordanlewis commented:
Ok, this was a *different* flake. We try again

bors r+

jordanlewis commented:
Yay

@rafiss we should retry some of the longer hibernate/django tests with this in, very curious to see if it makes a difference.
